### PR TITLE
Working to finish - feature/encore-webpack

### LINKED ISSUE #5454: [RFC] Use Webpack for asset management
## Feature Request

I would like to provide a PR which implement the usage [webpack encore](https://symfony.com/doc/current/frontend.html)

- [x] Setup webpack encore
- [x] install vendors via yarn (remove from `Resources/public/vendor`)
- [x] use custom stylesheets, javascript files also via webpack
- [x] remove bower

#### Vendors
- AdminLTE2
- famfamfam  (still needed?)
- jQuery
- jQuery form
- jQuery scrollTo
- jQuery UI
- mansonry
- readmoreJS
- slimScroll
- waypoints
- xEditable
- iCheck
- Bootstrap (corebundle)
- fontawesome (corebundle)
- ionicons (corebundle)
- datetimepicker (corebundle)
- select2 (corebundle)
- momentjs (corebundle)

#### Custom 
- colors.css
- layout.css
- styles.css
- tree.css
- jquery.confirmExit.js
- Admin.js
- sidebar.js
- treeview.js

### Benefits
- only one css, javascript file
- easier adding styles, javascripts
- usage of sass or less
- easy update of dependencies for example jQuery

### Questions
- which branch i should use master, 4.x, 3.x?

What would you say?

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

This PR rebase and prepare to merge work done in #5461 by @silasjoisten 

I am targeting this branch, because {reason}.

Closes #5454, #5461.

## Changelog

```markdown
### Added
- Added some `Class::newMethod()` to do great stuff.

### Changed

### Deprecated

### Removed

### Fixed

### Security
```

### CODE REVIEW

**Comment by wbloszyk** [comment]
This PR is not RTM but I want present some informations about it.
- 476 files will be remove from vendors
- yarn.lock with 7000+ lines will be added
- some files will be move or generated by webpack
- current AdminBundle assets system will be replace by encore-webpack

TODO:
- improve documentation
- check builds registred to twig global `sonata_admin_webpack_entries`
- remove `TODOs`

IMHO `encore-webpack` should be add to 3.x branch with some change:
- new config option `sonata_admin.options.assets_system` shoudl be added ['deprecated_assets', `encore-webpack`]
- vendors files should not be remove in 3.x
- `symfony/encore-webpack` should be suggested 

This will also allow add `Bootstrap 4` templates to `Sonata 3`.

WDYT? @greg0ire, @core23, @phansys, @jordisala1991, @VincentLanglet

**Comment by jordisala1991** [comment]
What about having 2 options: If I want to compile myself the assets using this way I can add them to my webpack encore configuration. But what about if I only want to have the assets compiled directly? Should we add here the minified one?

**Comment by VincentLanglet** [comment]
> What about having 2 options: If I want to compile myself the assets using this way I can add them to my webpack encore configuration. But what about if I only want to have the assets compiled directly? Should we add here the minified one?

I would prefer having the AdminBundle without the assets and create another repository with the minified assets.
This will avoid webpack users to download the minified assets for nothing. WDYT @jordisala1991 ?

**Comment by wbloszyk** [comment]
> What about having 2 options: If I want to compile myself the assets using this way I can add them to my webpack encore configuration. But what about if I only want to have the assets compiled directly? Should we add here the minified one?

Good idea. We should add webpack to sonata/sandbox, not only AdminBundle. (Probably only AdminBundle and PageBundle)

Step 1: 
Add encore-webpack as optional for people who know what to do with it 

Step 2:
Add `Boostrap 4` and `AdminLTE 3` templates

Step 3:
Prepare compiled assets and add it as another option

Step 4:
Remove `deprecated_assets` in master. People will be force to use webpack, anyway they got compiled assets, so knowlage about webpack will not be reqiure. 

@jordisala1991 This will be OK?

**Comment by wbloszyk** [comment]
> > What about having 2 options: If I want to compile myself the assets using this way I can add them to my webpack encore configuration. But what about if I only want to have the assets compiled directly? Should we add here the minified one?
> 
> I would prefer having the AdminBundle without the assets and create another repository with the minified assets.
> This will avoid webpack users to download the minified assets for nothing. WDYT @jordisala1991 ?

Something like symfony/recipies for assets?

**Comment by wbloszyk** [comment]
I start working on `Step 1`. Other things can be add later.

**Comment by jordisala1991** [comment]
This PR is more like: https://github.com/Sylius/Sylius (You have to build your own assets on the project using Sylius to have admin styles)
Static assets: https://github.com/EasyCorp/EasyAdminBundle (You have the webpack config there, but you have the compiled files generated on each release)

Maybe we can have the good things of both worlds.. that's what I was suggesting

File: docs/getting_started/installation.rst [hunk 1]
```diff
@@ -135,6 +135,22 @@ install the assets:
     bin/console cache:clear
     bin/console assets:install
 
+Webpack Encore Integration
+--------------------------
+
+When the assets are installed add following to  your configuration:
```
Comment (axzx) [inline docs/getting_started/installation.rst:141]:
```suggestion
When the assets are installed add following to your configuration:
```

**Comment by wbloszyk** [comment]
@jordisala1991 I check you proposition.

`Sylius/Sylius` is more like `sandbox`, where prople have shop and admin config.
`EasyAdminBundle` is more like `SonataAdminBundle` and it is only part of project.

I almost done alfa version for 3.x branch. I don't wanna release it becouse:

IMHO, better way will be add webpack to sandbox and then split them to sepete config/build. This way will be easier to get @jordisala1991 idea form  https://github.com/sonata-project/SonataAdminBundle/pull/6177#issuecomment-652514876

**Comment by wbloszyk** [comment]
Close in favor of https://github.com/sonata-project/dev-kit/issues/779

**Pull request closed without merging**